(window.webpackJsonp=window.webpackJsonp||[]).push([[106],{242:function(e,t,n){"use strict";n.r(t),n.d(t,"frontMatter",(function(){return i})),n.d(t,"metadata",(function(){return c})),n.d(t,"rightToc",(function(){return s})),n.d(t,"default",(function(){return l}));var r=n(2),a=n(9),o=(n(0),n(266)),i={title:"Exceptions"},c={id:"programming/Java/10Exception",isDocsHomePage:!1,title:"Exceptions",description:"Definition",source:"@site/docs/programming/Java/10Exception.mdx",permalink:"/docs/programming/Java/10Exception",editUrl:"https://github.com/lingt-xyz/io/tree/master/docs/programming/Java/10Exception.mdx",lastUpdatedAt:1598198409,sidebar:"java",previous:{title:"Inheritance",permalink:"/docs/programming/Java/9Inheritance"},next:{title:"Maven",permalink:"/docs/programming/Java/11Maven"}},s=[{value:"Definition",id:"definition",children:[]},{value:"The Three Kinds of Exceptions",id:"the-three-kinds-of-exceptions",children:[{value:"Checked Exception",id:"checked-exception",children:[]},{value:"Error",id:"error",children:[]},{value:"Runtime Exception",id:"runtime-exception",children:[]}]},{value:"Class Error",id:"class-error",children:[]},{value:"Class Exception",id:"class-exception",children:[]},{value:"Class RuntimeException",id:"class-runtimeexception",children:[]},{value:"Overriding",id:"overriding",children:[]}],p={rightToc:s};function l(e){var t=e.components,n=Object(a.a)(e,["components"]);return Object(o.b)("wrapper",Object(r.a)({},p,n,{components:t,mdxType:"MDXLayout"}),Object(o.b)("h2",{id:"definition"},"Definition"),Object(o.b)("blockquote",null,Object(o.b)("p",{parentName:"blockquote"},Object(o.b)("a",Object(r.a)({parentName:"p"},{href:"https://docs.oracle.com/javase/tutorial/essential/exceptions/definition.html"}),"An ",Object(o.b)("em",{parentName:"a"},"exception")," is an ",Object(o.b)("strong",{parentName:"a"},"event"),", which occurs during the execution of a program, that ",Object(o.b)("strong",{parentName:"a"},"disrupts")," the normal flow of the program's instructions."))),Object(o.b)("h2",{id:"the-three-kinds-of-exceptions"},"The Three Kinds of Exceptions"),Object(o.b)("h3",{id:"checked-exception"},"Checked Exception"),Object(o.b)("p",null,"These are exceptional conditions that a well-written application should ",Object(o.b)("strong",{parentName:"p"},"anticipate and recover from"),". For example, suppose an application prompts a user for an input file name, then opens the file by passing the name to the constructor for ",Object(o.b)("inlineCode",{parentName:"p"},"java.io.FileReader"),". Normally, the user provides the name of an existing, readable file, so the construction of the ",Object(o.b)("inlineCode",{parentName:"p"},"FileReader")," object succeeds, and the execution of the application proceeds normally. But sometimes the user supplies the name of a nonexistent file, and the constructor throws ",Object(o.b)("inlineCode",{parentName:"p"},"java.io.FileNotFoundException"),". A well-written program ",Object(o.b)("strong",{parentName:"p"},"will catch this exception and notify the user of the mistake, possibly prompting for a corrected file name"),"."),Object(o.b)("h3",{id:"error"},"Error"),Object(o.b)("p",null,"These are exceptional conditions that are ",Object(o.b)("em",{parentName:"p"},Object(o.b)("strong",{parentName:"em"},"external"))," to the application, and that the application usually ",Object(o.b)("em",{parentName:"p"},Object(o.b)("strong",{parentName:"em"},"cannot anticipate or recover from")),". For example, suppose that an application successfully opens a file for input, but is unable to read the file because of a hardware or system malfunction. The unsuccessful read will throw ",Object(o.b)("inlineCode",{parentName:"p"},"java.io.IOError"),". An application might choose to catch this exception, in order to notify the user of the problem \u2014 but it also might make sense for the program to print a stack trace and exit."),Object(o.b)("h3",{id:"runtime-exception"},"Runtime Exception"),Object(o.b)("p",null,"These are exceptional conditions that are ",Object(o.b)("em",{parentName:"p"},Object(o.b)("strong",{parentName:"em"},"internal"))," to the application, and that the application usually ",Object(o.b)("em",{parentName:"p"},Object(o.b)("strong",{parentName:"em"},"cannot anticipate or recover from")),". These usually indicate ",Object(o.b)("strong",{parentName:"p"},"programming bugs"),", such as logic errors or improper use of an API. For example, consider the application described previously that passes a file name to the constructor for ",Object(o.b)("inlineCode",{parentName:"p"},"FileReader"),". If a logic error causes a ",Object(o.b)("inlineCode",{parentName:"p"},"null")," to be passed to the constructor, the constructor will throw ",Object(o.b)("inlineCode",{parentName:"p"},"NullPointerException"),". The application can catch this exception, but it probably ",Object(o.b)("strong",{parentName:"p"},"makes more sense to eliminate the bug")," that caused the exception to occur."),Object(o.b)("p",null,'{% hint style="info" %}\nErrors and runtime exceptions are collectively known as unchecked exceptions.\n{% endhint %}'),Object(o.b)("h2",{id:"class-error"},"Class Error"),Object(o.b)("p",null,"An ",Object(o.b)("inlineCode",{parentName:"p"},"Error")," is a subclass of ",Object(o.b)("inlineCode",{parentName:"p"},"Throwable")," that indicates serious problems that a reasonable application ",Object(o.b)("strong",{parentName:"p"},"should not try to catch"),". Most such errors are ",Object(o.b)("strong",{parentName:"p"},"abnormal")," conditions. "),Object(o.b)("p",null,"A method is not required to declare in its ",Object(o.b)("inlineCode",{parentName:"p"},"throws")," clause any subclasses of ",Object(o.b)("inlineCode",{parentName:"p"},"Error")," that might be thrown during the execution of the method but not caught, since these errors are abnormal conditions that should never occur. That is, ",Object(o.b)("inlineCode",{parentName:"p"},"Error")," and its subclasses are regarded as ",Object(o.b)("em",{parentName:"p"},Object(o.b)("strong",{parentName:"em"},"unchecked exceptions"))," for the purposes of compile-time checking of exceptions"),Object(o.b)("h2",{id:"class-exception"},"Class Exception"),Object(o.b)("p",null,"The class ",Object(o.b)("inlineCode",{parentName:"p"},"Exception")," and its subclasses are a form of ",Object(o.b)("inlineCode",{parentName:"p"},"Throwable")," that indicates conditions that a reasonable application might want to catch."),Object(o.b)("p",null,"The class ",Object(o.b)("inlineCode",{parentName:"p"},"Exception")," and any subclasses that are not also subclasses of ",Object(o.b)("a",Object(r.a)({parentName:"p"},{href:"https://docs.oracle.com/javase/8/docs/api/java/lang/RuntimeException.html"}),Object(o.b)("inlineCode",{parentName:"a"},"RuntimeException"))," are ",Object(o.b)("em",{parentName:"p"},Object(o.b)("strong",{parentName:"em"},"checked exceptions")),". Checked exceptions need to be declared in a method or constructor's ",Object(o.b)("inlineCode",{parentName:"p"},"throws")," clause if they can be thrown by the execution of the method or constructor and propagate outside the method or constructor boundary."),Object(o.b)("h2",{id:"class-runtimeexception"},"Class RuntimeException"),Object(o.b)("p",null,Object(o.b)("inlineCode",{parentName:"p"},"RuntimeException")," is the superclass of those exceptions that can be thrown during the normal operation of the Java Virtual Machine."),Object(o.b)("p",null,Object(o.b)("inlineCode",{parentName:"p"},"RuntimeException")," and its subclasses are ",Object(o.b)("em",{parentName:"p"},Object(o.b)("strong",{parentName:"em"},"unchecked exceptions")),". Unchecked exceptions do ",Object(o.b)("em",{parentName:"p"},"not")," need to be declared in a method or constructor's ",Object(o.b)("inlineCode",{parentName:"p"},"throws")," clause if they can be thrown by the execution of the method or constructor and propagate outside the method or constructor boundary."),Object(o.b)("p",null,Object(o.b)("img",Object(r.a)({parentName:"p"},{src:".gitbook/assets/exception_classes.png",alt:"https://www3.ntu.edu.sg/home/ehchua/programming/java/J5a_ExceptionAssert.html"}))),Object(o.b)("h2",{id:"overriding"},"Overriding"),Object(o.b)("p",null,"An overriding method cannot declare exception types that were not declared in its original. However, it may declare exception types are the same as, or subclass of its original. It needs not declare all the exceptions as its original. It can throw ",Object(o.b)("strong",{parentName:"p"},"fewer")," exceptions than the original, ",Object(o.b)("strong",{parentName:"p"},"but not more"),"."))}l.isMDXComponent=!0},266:function(e,t,n){"use strict";n.d(t,"a",(function(){return b})),n.d(t,"b",(function(){return d}));var r=n(0),a=n.n(r);function o(e,t,n){return t in e?Object.defineProperty(e,t,{value:n,enumerable:!0,configurable:!0,writable:!0}):e[t]=n,e}function i(e,t){var n=Object.keys(e);if(Object.getOwnPropertySymbols){var r=Object.getOwnPropertySymbols(e);t&&(r=r.filter((function(t){return Object.getOwnPropertyDescriptor(e,t).enumerable}))),n.push.apply(n,r)}return n}function c(e){for(var t=1;t<arguments.length;t++){var n=null!=arguments[t]?arguments[t]:{};t%2?i(Object(n),!0).forEach((function(t){o(e,t,n[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(n)):i(Object(n)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(n,t))}))}return e}function s(e,t){if(null==e)return{};var n,r,a=function(e,t){if(null==e)return{};var n,r,a={},o=Object.keys(e);for(r=0;r<o.length;r++)n=o[r],t.indexOf(n)>=0||(a[n]=e[n]);return a}(e,t);if(Object.getOwnPropertySymbols){var o=Object.getOwnPropertySymbols(e);for(r=0;r<o.length;r++)n=o[r],t.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(e,n)&&(a[n]=e[n])}return a}var p=a.a.createContext({}),l=function(e){var t=a.a.useContext(p),n=t;return e&&(n="function"==typeof e?e(t):c(c({},t),e)),n},b=function(e){var t=l(e.components);return a.a.createElement(p.Provider,{value:t},e.children)},u={inlineCode:"code",wrapper:function(e){var t=e.children;return a.a.createElement(a.a.Fragment,{},t)}},m=a.a.forwardRef((function(e,t){var n=e.components,r=e.mdxType,o=e.originalType,i=e.parentName,p=s(e,["components","mdxType","originalType","parentName"]),b=l(n),m=r,d=b["".concat(i,".").concat(m)]||b[m]||u[m]||o;return n?a.a.createElement(d,c(c({ref:t},p),{},{components:n})):a.a.createElement(d,c({ref:t},p))}));function d(e,t){var n=arguments,r=t&&t.mdxType;if("string"==typeof e||r){var o=n.length,i=new Array(o);i[0]=m;var c={};for(var s in t)hasOwnProperty.call(t,s)&&(c[s]=t[s]);c.originalType=e,c.mdxType="string"==typeof e?e:r,i[1]=c;for(var p=2;p<o;p++)i[p]=n[p];return a.a.createElement.apply(null,i)}return a.a.createElement.apply(null,n)}m.displayName="MDXCreateElement"}}]);